File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
class Solution {
    public int shipWithinDays(int[] weights, int days) {
        int lCapacity = 0;
        int uCapacity = 0;
        int result = 0;

        for(int i : weights) {
            lCapacity = Math.max(lCapacity,i);
            uCapacity += i;
        }
        if(days == 1) {
            return uCapacity;
        }

        while(lCapacity < uCapacity) {
            int mid = lCapacity + (uCapacity - lCapacity) / 2;
            if(!canBeShipped(weights,days,mid)) {
                lCapacity = mid + 1;
            } else {
                result = mid;
                uCapacity = mid;
            }
        }

        return result;
    }

    public boolean canBeShipped(int[] weights, int days, int k) {
        int totalDays = 0, weight = 0, i = 0;
        while (i < weights.length) {
            int j = i;
            while(j < weights.length && weight <= k) {
                weight += weights[j];
                if(weight > k) {
                    i = j;
                    break;
                }
                j++;
            }
            weight = 0;
            totalDays++;
            i = (j == weights.length) ? j : i;
        }
        return totalDays <= days;
    }

}

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
class Solution(object):
    def shipWithinDays(self, weights, days):
        least_capacity = max(weights)
        highest_capacity = sum(weights)

        start, end = least_capacity, highest_capacity

        result = 0
        while start < end:
            mid = start + (end - start) / 2
            if self.isCapacityWithinDays(weights,mid,days):
                end = mid
            else:
                start = mid + 1
        
        return start

    
    def isCapacityWithinDays(self,weights,capacity, days):
        numberOfDays = 1
        total_sum = 0
        for i in weights:
            total_sum += i
            if total_sum > capacity:
                numberOfDays += 1
                total_sum = i
            
        return numberOfDays <= days

        